File: P008/Python/timeSeries_AOA.py
#!/usr/bin/env python

#-------------
# Load modules
#-------------
import os
import sys
import time
import datetime as dt
from netCDF4 import Dataset
import numpy as np
import glob
#import matplotlib.pyplot as plt            # pyplot module import
#from matplotlib.ticker import ScalarFormatter, FormatStrFormatter
import vertLevels_GEOS5 as pressLevels # local module

def is_leap_year(year):
    """Determine whether a year is a leap year."""
    return year % 4 == 0 and (year % 100 != 0 or year % 400 == 0)

print "-----------------------------"
print "Serial Time Series Processing"
print "-----------------------------"

# Start time
#-----------
start = time.clock()
begTime = dt.datetime.now()

vName = 'aoa'
begYear = 1990
endYear = 2009

firstFile = 0
ref_lat = -86.0

numDays = 0

coef = 365.5

refDIR = '/discover/nobackup/jkouatch/GEOS_CTM/ProductionRuns/runAOA/outFiles/TR/'

# Loop over the years
#--------------------
for year in range(begYear, endYear+1):
    #print "   Processing files for ", year
    dirname = 'Y'+str(year)
    dirY = os.path.join(refDIR,dirname)

    listFiles = glob.glob(dirY+"/runAOA.TR."+str(year)+"*_1200z.nc4")

    numDays += len(listFiles)
    #print numDays

    # Loop over the daily files
    #--------------------------
    for file in listFiles:
        #numDays += 1
        # Open file
        #----------
        nf = Dataset(file, mode='r')

        # Extract information if it is the first file
        #--------------------------------------------
        if firstFile == 0:
           #firstFile = 1
           lons = nf.variables['lon'][:]
           lats = nf.variables['lat'][:]
           levs = nf.variables['lev'][::-1]
           nlons = np.size(lons)
           nlats = np.size(lats)
           nlevs = np.size(levs)

           levs = pressLevels.calcPressureLevels(nlevs)
           lat_index = (np.abs(lats - ref_lat)).argmin()
           #print "lat_index: ", lat_index

           #dataVal   = np.zeros((nlevs))

        # Read the daily average age of air
        #----------------------------------
        var = nf.variables[vName][0,::-1,lat_index,:] / coef

        # Determine the zonal mean
        #-------------------------
        tempVar = np.mean(var,axis=1)

        # Stack the daily values into an existing array
        #----------------------------------------------
        if firstFile == 0:
           firstFile = 1
           dataVal = tempVar
        else:
           dataVal = np.column_stack((dataVal,tempVar))

        #print dataVal.shape

        # Close file
        #-----------
        nf.close()

# End time
#-----------
endTime = dt.datetime.now()
diffTime = endTime - begTime
#finish = time.clock()

print
#print 'Elapsed Time for the time series calculations: ', finish - start,'s'
print 'Elapsed Time for the time series calculations: ', diffTime.total_seconds(),'s'
print

# Verification
# Make sure that the number of files read is 7305
#-------------
if (numDays == 7305):
   print "Successful reading"
else:
   print "Reading as not succcessful"

#sys.exit()

#days = [i+1 for i in range(numDays)]


# Plot the mean at a specified level
#-----------------------------------
#figName  = 'fig_TimeSeries_AgeOfAir'

#X, Y = np.meshgrid(np.array(days), levs)

#fig = plt.figure(figsize=(13,13))

#ax = fig.add_subplot(1,1,1)

#CS = ax.contour(X, Y, dataVal)
#ax.clabel(CS, inline=1, fontsize=10)
#ax.set_title('Age-of-Air (years) at  $86^o$  S')
## y-axis
#ax.set_yscale('log')
#ax.set_ylim(plt.ylim()[::-1])
#ax.set_ylabel('Pressure (hPa)')
#ax.yaxis.set_major_formatter(FormatStrFormatter('%.2f'))
#
## x-axis
#x_ticks = [day for day in days if day%365 == 0] # Only pull out full years
#x_labels = [str(i+begYear) for i in range(len(x_ticks))]
#ax.set_xticks(x_ticks)
#ax.set_xticklabels(x_labels)
#ax.set_xlabel('Year')
#ax.xaxis.grid(True)
#

#plt.savefig(figName+'.png')

#plt.show()

File: P008/Julia/timeSeries_AOA.jl
#!/usr/bin/env julia

#-------------
# Load modules
#-------------
using NetCDF

function calcPressureLevels(nlevs)
    """
      This function takes the number of vertical levels
      to read a file that contains the values of ak and bk.
      It then computes the pressure levels using the formula:

          0.5*(ak[l]+ak[l+1]) + 0.1*1.0e5*(bk[l]+bk[l+1])

      Input Varialble:
        nlevs: number of vertical levels

      Returned Value:
        phPa: pressure levels from bottom to top
    """
    ak   = zeros(Float32, nlevs+1)
    bk   = zeros(Float32, nlevs+1)
    phPa = zeros(Float32, nlevs)

    fileName = string(string(nlevs), "-layer.p")

    fid = open(fileName) ;
    lines = readlines(fid) ;
    lnum = 1 ;
    k = 1 ;
    for line in lines
        lnum += 1 ;
        if (lnum > 3)
           line    = strip(line) ;
           columns = split(line) ;
           ak[k]   = float(columns[2]) ;
           bk[k]   = float(columns[3]) ;
           k      += 1 ;
        end
    end
    close(fid);

    for k in 1:nlevs
        phPa[k] = 0.50*((ak[k]+ak[k+1])+0.01*1.00e+05*(bk[k]+bk[k+1]));
    end

    return phPa[nlevs:-1:1]
end


function getData(file)
   coef = 365.5
   vName = "aoa"
   ref_lat = -86.0

   #lats = ncread(file, "lat")
   #lat_index = findfirst(lats, ref_lat)
   #startList = [1,lat_index,1,1]
   startList = [1,5,1,1]
   countList = [-1,1,-1,1]

   # Read the daily average age of air
   var = ncread(file, vName, start=startList, count=countList) / coef

   # Determine the zonal mean
   tempVar = mean(var,1)
   tempVar = tempVar[1,1,:,1]

   return tempVar
end

function getLevels(file)
   levs = ncread(file, "lev")
   nlevs = length(levs)
   levs = calcPressureLevels(nlevs)
   return levs
end 

# Start time
#-----------
tic()

begYear = 1990
endYear = 2009

firstFile = 0

numDays = 0

dataVal = zeros(72)

refDIR = "/discover/nobackup/jkouatch/GEOS_CTM/ProductionRuns/runAOA/outFiles/TR/"

# Loop over the years
#--------------------
for year in begYear:endYear
    #println( "Processing files for ", year)
    dirY = refDIR*"Y"* string(year)*"/"

    listFiles = filter(x -> endswith(x, "_1200z.nc4"),  readdir(dirY))
      
    listFiles = [file for file in listFiles]

    numDays = numDays + length(listFiles)

    # Loop over the daily files
    #--------------------------
    for fl in listFiles
        file = string(dirY, fl)

        #println( "     --->File: ", file)
        tempVar = getData(file)

        if (firstFile == 0)
           levs = getLevels(file)
        end 
        # Stack the daily values into an existing array
        #----------------------------------------------
        dataVal = hcat(dataVal,tempVar)
        #println("       Size of dataVal: ", size(dataVal))
    
    end
end

# Remove the first row
#---------------------
dataVal = dataVal[setdiff(1:end, 1),:] # delete the first row

# End time
#-----------
toc()

if (numDays == 7305)
   println("Successfully read ", numDays, "files")
else
   println("Only read ", numDays, "files! Check your script.")
end
